Simulated Inertial Sensors in C++

psiorx opened the pull request:
Re-opened version.

Implements:
 NoiseModel interface
 AdditiveGaussianNoise
 RigidBodyAccelerometer
 RigidBodyGyroscope

This also adds unit tests for these sensors and includes usage in the quadrotor LCM output example.

psiorx commented:
Adding a `RigidBodyGyroscope` unit test and I'll zap `std::enable_shared_from_this` in favor of references.  Stand by.

RussTedrake commented:
the travis build looks like a legitimate error in testAccelerometer?

RussTedrake commented on drake/examples/Quadrotor/QuadrotorControl.h:
```diff
@@ -10,7 +10,9 @@ class QuadrotorControl {
 
-    QuadrotorControl(void) : motors(Eigen::Vector4d::Zero()) {}
+    QuadrotorControl(void) : motors(Eigen::Vector4d::Zero()) {
+      motors << 1.226250, 1.226250, 1.226250, 1.226250;
```
would it be cleaner to pass in the mass (or the quadrotor plant) and use mass/4 here?

RussTedrake commented on drake/examples/Quadrotor/QuadrotorState.h:
```diff
@@ -8,47 +8,82 @@ template <typename ScalarType = double>
+        orientation.setZero();
+        twist.setZero();
+        accelerometer.setZero();
+        gyroscope.setZero();
```
this isn't the quadrotor state anymore.  it's quadrotor output.?

RussTedrake commented on drake/examples/Quadrotor/runDynamics.cpp:
```diff
@@ -18,8 +18,20 @@ int main(int argc, char* argv[]) {
   auto const & tree = rigid_body_sys->getRigidBodyTree();
+  rigid_body_sys->addRobotFromFile(getDrakePath()+"/examples/Quadrotor/quadrotor.urdf", floating_base_type);
+
+  auto sensor_frame = tree->frames[0]; //this is a propellor frame.  probably want a body frame for this
```
can you look it up by name, instead of hard-coding the 0 index?

RussTedrake commented on drake/lcmtypes/lcmt_quadrotor_state_t.lcm:
```diff
@@ -3,6 +3,10 @@ package drake;
 {
   int64_t timestamp;
-  double x[13];
+  double position[3];
```
again... this is now the quadrotor_output, not the state.

RussTedrake commented on drake/systems/plants/RigidBodySystem.cpp:
```diff
@@ -367,8 +361,53 @@ RigidBodySpringDamper::RigidBodySpringDamper(RigidBodySystem* sys,
+  auto const& q = rigid_body_state.getQ();
+  auto const& v = rigid_body_state.getV();
+  VectorXd x(q.size() + v.size());
+  x << q, v;
```
we should have a rigid_body_state.getX() so you don't have to do this everywhere.  (i didn't actually know we had getQ and getV; there are probably places where I should have used those!)

RussTedrake commented on drake/systems/plants/RigidBodySystem.cpp:
```diff
@@ -198,22 +194,19 @@ RigidBodySystem::StateVector<double> RigidBodySystem::dynamics(
-    return y;
+RigidBodySystem::OutputVector<double> RigidBodySystem::output(const double& t,
+                                                              const RigidBodySystem::StateVector<double>& x,
+                                                              const RigidBodySystem::InputVector<double>& u) const
```
i don't see anywhere where you set the direct feedthrough flag to true if you have an accelerometer.

RussTedrake commented on drake/systems/plants/RigidBodySystem.cpp:
```diff
@@ -367,8 +361,53 @@ RigidBodySpringDamper::RigidBodySpringDamper(RigidBodySystem* sys,
+  Vector3d accel_base = Jdot_times_v + J * v_dot;
+  Vector3d accel_body = quatRotateVec(quat_world_to_body, accel_base);
+  
+  return noise_model ? noise_model->generateNoise(accel_body) : accel_body;
```
i think we're going to want to handle this differently in the long run (so the analysis can go through without noise in a generic way for all systems, etc).  but i think it's fine for now.

RussTedrake commented:
really nice work, @psiorx .  thank you for doing this!!

psiorx commented:
@RussTedrake 
Thanks for the review.  I've addressed all your comments.  

The testAccelerometer failure was indeed real (I was using an Eigen block operation on a temporary).

```C++```
sys->output(t, x, u).tail<3>();
```
This attempts to read invalid memory as the temporary is not guaranteed to exist at the time the expression is evaluated.

I love being able to use Valgrind on Drake code now.  I've already caught 3 bugs in the last few days with it.  We should consider running some of our unit tests through it on the build servers and checking for memory errors.

david-german-tri commented:
@psiorx, it possible this PR broke the Win32 build?

https://drake-jenkins.csail.mit.edu/job/windows-msvc-32-continuous/40/

(Search for "pendulumURDFDynamicsTest" in the console log.)

RussTedrake commented:
yes.  i think it probably did (sorry for not catching it... it was masked by the bullet issues).  
this is another case where make_shared is failing to use the eigen aligned_allocator.  have to go back to new shared_ptr or allocate_shared.
@psiorx -- are you on it?

psiorx commented:
Looking into it.
On Mar 8, 2016 9:32 PM, "Russ Tedrake" <notifications@github.com> wrote:

> yes. i think it probably did (sorry for not catching it... it was masked
> by the bullet issues).
>
> this is another case where make_shared is failing to use the eigen
> aligned_allocator. have to go back to new shared_ptr or allocate_shared.
> @psiorx <https://github.com/psiorx> -- are you on it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RobotLocomotion/drake/pull/1814#issuecomment-194077524>
> .
>

psiorx commented:
@RussTedrake Having trouble debugging this on the build servers.

I keep getting a `fatal error C1060: compiler is out of heap space` when running `make`.
